### PR TITLE
add graceful shutdown

### PR DESCRIPTION
**Pull request by silesky**
### Avoid losing events on exit!
 * Call `.closeAndFlush()` to stop collecting new events and flush all existing events.
  * If a callback on an event call is included, this also waits for all callbacks to be called, and any of their subsequent promises to be resolved.
```ts
await analytics.closeAndFlush()
// or
await analytics.closeAndFlush({ timeout: 5000 }) // force resolve after 5000ms
```
### Graceful Shutdown: Advanced Example
```ts
import express from 'express'
const app = express()

const server = app.listen(3000)
app.get('/', (req, res) => res.send('Hello World!'));

const onExit = async () => {
   console.log("Gracefully closing server...");
   await analytics.closeAndFlush() // flush all existing events
   server.close(() => process.exit());
 };

process.on("SIGINT", onExit);
process.on("SIGTERM", onExit);


```
https://github.com/segmentio/analytics-next/blob/9d021e2f3ea78099dd12ec50622483f945267250/packages/node/README.md

### CODE REVIEW

File: packages/node/src/app/analytics-node.ts [hunk 1]
```diff
@@ -103,11 +102,48 @@ export class AnalyticsNode
...
+      } else {
+        this.once('drained', () => resolve())
+      }
+    })
+    return timeout ? pTimeout(promise, timeout).catch(() => undefined) : promise
+  }
+
   private _dispatch(segmentEvent: CoreSegmentEvent, callback?: Callback) {
+    if (this._isClosed) {
+      return undefined
```
[the remark this answers is not in the document]
Comment (silesky) [reply packages/node/src/app/analytics-node.ts:129]: I think we could emit an event like "denied" here. That way,  a user could collect / log them and even use them to re-hydrate (if they cared to)

File: packages/core/src/callback/index.ts [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
+      .then((val) => {
+        clearTimeout(timeoutId)
+        return resolve(val)
+      })
+      .catch(reject)
   })
 }
 
-function sleep(timeoutInMs: number): Promise<void> {
+export function sleep(timeoutInMs: number): Promise<void> {
```
[the remark this answers is not in the document]
Comment (silesky) [reply packages/core/src/callback/index.ts:19]: yes, we define sleep everywhere!

File: packages/node/src/app/analytics-node.ts [hunk 3]
```diff
@@ -103,11 +102,48 @@ export class AnalyticsNode
...
+  /**
+   * Call this method to stop collecting new events and flush all existing events.
+   * This method also waits for any event method-specific callbacks to be triggered,
+   * and any of their subsequent promises to be resolved/rejected.
+   */
+  public closeAndFlush({
+    timeout,
+  }: {
+    /** Set a maximum time permitted to wait before resolving. Default = no maximum. */
+    timeout?: number
```
[the remark this answers is not in the document]
Comment (silesky) [reply packages/node/src/app/analytics-node.ts:114]: Hmmm, If we do force resolve and there are pending promises (like HTTP Requests), they wouldn't be promises whose resolution was blocking an end-user network request. Those unresolved promise callbacks would just hang out on the job queue which will eventually get torn down. AFAIK, unresolved setTimeouts wouldn't effect successful shutdown either way -- express servers just care about closing sockets (and even then, not that much).

File: packages/core/src/callback/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisradek) [reply packages/core/src/callback/index.ts:19]: That sounds good to me!

**Review by chrisradek — APPROVED** [review]
Looks good to me, :shipit:

**Pull request merged**